ownCloud/Gallery/GalleryHostViewController.swift

```diff
@@ -0,0 +1,229 @@
+			return DisplayViewController()
+		}
+
+		return controllerType
+	}
+
+	// MARK: - Helper methods
+	private func viewControllerAtIndex(index: Int) -> UIViewController? {
+		guard let items = items else { return nil }
+
+		guard index >= 0, index < items.count else { return nil }
+
+		let item = items[index]
+
+		let newViewController = selectDisplayViewControllerBasedOn(mimeType: item.mimeType!)
+		let configuration = configurationFor(item, viewController: newViewController)
+
+		newViewController.configure(configuration)
+		newViewController.present(item: item)
+		return newViewController
+	}
+
+	private func configurationFor(_ item: OCItem, viewController: UIViewController) -> DisplayViewConfiguration {
+		let shouldDownload = viewController is (DisplayViewController & DisplayExtension) ? true : false
+		var configuration: DisplayViewConfiguration
+		if !shouldDownload {
+			configuration = DisplayViewConfiguration(item: item, core: core, state: .notSupportedMimeType)
+		} else {
+			configuration = DisplayViewConfiguration(item: item, core: core, state: .hasNetworkConnection)
+		}
+		return configuration
+	}
+}
+
+extension GalleryHostViewController: UIPageViewControllerDataSource {
+	func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
+		if let displayViewController = viewControllers?.first as? DisplayViewController,
+			let item = displayViewController.item,
+			let index = items?.firstIndex(where: {$0.fileID == item.fileID}) {
+			return viewControllerAtIndex(index: index + 1)
+		}
+
+		return nil
+
+	}
+
+	func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
+
+		if let displayViewController = viewControllers?.first as? DisplayViewController,
+			let item = displayViewController.item,
+			let index = items?.firstIndex(where: {$0.fileID == item.fileID}) {
+			return viewControllerAtIndex(index: index - 1)
+		}
+
+		return nil
+	}
+}
+
+extension GalleryHostViewController: UIPageViewControllerDelegate {
+	func pageViewController(_ pageViewController: UIPageViewController, didFinishAnimating finished: Bool, previousViewControllers: [UIViewController], transitionCompleted completed: Bool) {
+
+		let previousViewController = previousViewControllers[0]
+		previousViewController.didMove(toParent: nil)
+
+		if completed, let viewControllerToTransition = self.viewControllerToTansition {
+			if self.children.contains(viewControllerToTransition) == false {
+				self.addChild(viewControllerToTransition)
+			}
+			viewControllerToTransition.didMove(toParent: self)
+			viewControllerToTransition.updateNavigationBarItems()
+		}
+	}
+
+	func pageViewController(_ pageViewController: UIPageViewController, willTransitionTo pendingViewControllers: [UIViewController]) {
+		if let viewControllerToTransition = pendingViewControllers[0] as? DisplayViewController {
```

Comment:
@pablocarmu is it ensured that `pendingViewControllers` is always not empty?

Reply:
coded a check for this